blog: add post about OVN-Kubernetes multi-net policies (#911)
kubevirt/kubevirt.github.io: _posts/2023-07-10-OVN-kubernetes-secondary-networks-policies.md

description: This post explains how to configure NetworkPolicies for KubeVirt VMs secondary networks.
navbar_active: Blogs
pub-date: July 10
pub-year: 2023
category: news
tags:
  [
    "Kubevirt",
    "kubernetes",
    "virtual machine",
    "VM",
    "SDN",
    "OVN",
    "NetworkPolicy"
  ]
comments: true
---

## Introduction
Kubernetes `NetworkPolicies` are constructs to control traffic flow at the IP
address or port level (OSI layers 3 or 4).
They allow the user to specify how a pod (or group of pods) is allowed to
communicate with other entities on the network. In simpler words: the user can
specify ingress from or egress to other workloads, using L3 / L4 semantics.

Keeping in mind `NetworkPolicy` is a Kubernetes construct - which only cares
about a single network interface - they are only usable for the cluster's
default network interface. This leaves a considerable gap for Virtual Machine
users, since they are heavily invested in secondary networks.

The k8snetworkplumbingwg has addressed this limitation by providing a
`MultiNetworkPolicy` CRD - it features the exact same API as `NetworkPolicy`
but can target `network-attachment-definitions`.

> OVN-Kubernetes implements this API, and configures access control accordingly

Reviewer: ```suggestion
[OVN-Kubernetes](https://github.com/ovn-org/ovn-kubernetes) implements this API, and configures access control accordingly
```

for secondary networks in the cluster.

In this post we will see how we can govern access control for VMs using the
multi-network policy API. On our simple example, we'll only allow into our VMs
for traffic ingressing from a particular CIDR range.

## Current limitations of `MultiNetworkPolicies` for VMs
Kubernetes `NetworkPolicy` has three types of policy peers:
- namespace selectors: allows ingress-from, egress-to based on the peer's namespace
- pod selectors: allows ingress-from, egress-to based on the peer's labels
- ip block: allows ingress-from, egress-to based on the peer's IP address

While `MultiNetworkPolicy` allows these three types, when used with VMs we
recommend using **only** the `IPBlock` policy peer - both `namespace` and `pod`

> selectors prevent the live-migration of Virtual Machines.

Reviewer: Could you please refer the reader to a source or give a note about why the usage of the above selectors prevents live migrations?

Author: Is the provided explanation good enough, or do you think it should be more detailed ?

Reviewer: Thank you for adding the explanation.
I think its enough.

> ## Demo

Reviewer: Could you please give a high-level description of what the setup will look like?
A block diagram could be nice.

Author: I'm stating below this cluster is composed of one control plane node plus two worker nodes.

What more info are you expecting to have ?

Reviewer: You can resolve this thread.
The information you've added below is enough.

Author: OK, thanks.

To run this demo, we will prepare a Kubernetes cluster with the following
components installed:
- [OVN-Kubernetes](https://github.com/ovn-org/ovn-kubernetes)
- [multus-cni](https://github.com/k8snetworkplumbingwg/multus-cni)
- [KubeVirt](https://github.com/kubevirt/kubevirt)
- [Multi-Network policy API](https://github.com/k8snetworkplumbingwg/multi-networkpolicy)

> The [following section](#environment-setup) will show you how to create a

Reviewer: I think the `#environment-setup` link is broken.

Author: Good catch.

Done.

[KinD](https://kind.sigs.k8s.io/) cluster, with upstream latest OVN-Kubernetes,
and upstream latest multus-cni deployed.

### Setup demo environment
Refer to the OVN-Kubernetes repo
[KIND documentation](https://github.com/ovn-org/ovn-kubernetes/blob/master/docs/kind.md#ovn-kubernetes-kind-setup)
for more details; the gist of it is you should clone the OVN-Kubernetes
repository, and run their kind helper script:

```bash
git clone git@github.com:ovn-org/ovn-kubernetes.git

cd ovn-kubernetes
pushd contrib ; ./kind.sh --multi-network-enable ; popd
```

> This will get you a running kind cluster, configured to use OVN-Kubernetes as

Reviewer: How many nodes will this cluster have?

Author: Done.

the default cluster network, configuring the multi-homing OVN-Kubernetes feature
gate, and deploying

> [multus-cni](https://github.com/k8snetworkplumbingwg/multus-cni) in the cluster.

Reviewer: nit: There is already a link to multus above.
Please consider removing it.

Author: Done.

### Install KubeVirt in the cluster
Follow Kubevirt's
[user guide](https://kubevirt.io/user-guide/operations/installation/#installing-kubevirt-on-kubernetes)

> to install the latest released version (currently, v0.59.2).

Reviewer: ```suggestion
to install the latest released version (currently, v1.0.0).
```

```bash
export RELEASE=$(curl https://storage.googleapis.com/kubevirt-prow/release/kubevirt/kubevirt/stable.txt)
kubectl apply -f "https://github.com/kubevirt/kubevirt/releases/download/${RELEASE}/kubevirt-operator.yaml"
kubectl apply -f "https://github.com/kubevirt/kubevirt/releases/download/${RELEASE}/kubevirt-cr.yaml"
kubectl -n kubevirt wait kv kubevirt --timeout=360s --for condition=Available
```

Now we have a Kubernetes cluster with all the pieces to start the Demo.

### Limiting ingress to a KubeVirt VM
In this example, we will configure a `MultiNetworkPolicy` allowing ingress into
our VMs only from a particular CIDR range - let's say `10.200.0.0/30`.

Provision the following NAD (to allow our VMs to live-migrate, we do not define
a `subnet`):
```yaml
---
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  name: flatl2net
spec:

>   config: |2

Reviewer: nit: Is this 2 intentional?
If so, what does it mean?

Reviewer: Thank you for the explenation.

    {
            "cniVersion": "0.4.0",
            "name": "flatl2net",
            "type": "ovn-k8s-cni-overlay",
            "topology":"layer2",
            "netAttachDefName": "default/flatl2net"
    }
```

> Let's now provision our VMs.

Reviewer: Please consider moving the table bellow to here, so the reader will understand what are the differences between all of the following VMs, without going over their spec.

Its also worth mentioning IMO that all of the entities (NAD and VMs) are at the `default` namespace.

Author: Done.

        name: flatl2-overlay
      termination/GracePeriodSeconds: 30
      volumes:
      - containerDisk:
          image: quay.io/kubevirt/fedora-with-test-tooling-container-disk:v1.0.0
        name: containerdisk
      - cloudInitNoCloud:
          networkData: |
            ethernets:
              eth0:
                addresses:
                - 10.200.0.1/24
            version: 2
          userData: |-
            #cloud-config
            user: fedora
            password: password
            chpasswd: { expire: False }
        name: cloudinitdisk
---
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  labels:
    kubevirt.io/vm: vm2
  name: vm2
spec:
  running: true
  template:
    metadata:
      labels:
        name: access-control
        kubevirt.io/domain: vm2
        kubevirt.io/vm: vm2
    spec:
      domain:
        devices:
          disks:
          - disk:
              bus: virtio
            name: containerdisk
          - disk:
              bus: virtio
            name: cloudinitdisk
          interfaces:
          - bridge: {}
            name: flatl2-overlay
          rng: {}
        resources:
          requests:

>             memory: 1024Mi

Reviewer: nit: According to @enp0s3, its better to use the `guest` field of http://kubevirt.io/api-reference/v1.0.0/definitions.html#_v1_memory to require memory for the guest.

Author: ignoring this nit since it's not relevant.

            name: cloudinitdisk
          interfaces:
          - bridge: {}
            name: flatl2-overlay
          rng: {}
        resources:
          requests:
            memory: 1024Mi
      networks:
      - multus:
          networkName: flatl2net
        name: flatl2-overlay
      termination/GracePeriodSeconds: 30
      volumes:
      - containerDisk:
          image: quay.io/kubevirt/fedora-with-test-tooling-container-disk:v1.0.0
        name: containerdisk
      - cloudInitNoCloud:
          networkData: |
            ethernets:
              eth0:
                addresses:
                - 10.200.0.6/24
            version: 2
          userData: |-
            #cloud-config
            user: fedora
            password: password
            chpasswd: { expire: False }
        name: cloudinitdisk
```

After this step, we should have the following deployment:

![image](/assets/2023-07-10-OVN-kubernetes-secondary-networks-policies/01-vms-provisioned.png)

Meaning, six VMs, with the following name to IP address (statically configured
via cloud-init) association:
- vm1: 10.200.0.1
- vm2: 10.200.0.2
- vm3: 10.200.0.3
- vm4: 10.200.0.4
- vm5: 10.200.0.5

> - vm6: 10.200.0.6

Reviewer: Please consider moving this part before the VM specs.

Author: Done.

> 

Reviewer: Please consider adding a sentence regarding the fact that you are about to connect to the VMs' serial console using the virtctl utility (with a link to its documentation).

Author: Done.

64 bytes from 10.200.0.6: icmp_seq=4 ttl=64 time=0.970 ms

--- 10.200.0.6 ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 3005ms
rtt min/avg/max/mdev = 0.970/1.275/1.850/0.350 ms
```

Let's now provision our `MultiNetworkPolicy`.
```bash
---
apiVersion: k8s.cni.cncf.io/v1beta1
kind: MultiNetworkPolicy
metadata:
  name:  ingress-ipblock
  annotations:
    k8s.v1.cni.cncf.io/policy-for: default/flatl2net
spec:
  podSelector:
    matchLabels:
        name: access-control
  policyTypes:
  - Ingress
  ingress:
  - from:
    - ipBlock:
        cidr: 10.200.0.0/30
```

The policy above will apply for all pods attached to the `flatl2net` secondary
network in the `default` namespace (from the `k8s.v1.cni.cncf.io/policy-for`
annotation) having the `access-control` label (from the network policy pod
selector `matchLabels` stanza).

Once the aforementioned policy is provisioned, ingress into these
`access-control` labeled pods will only be accepted if the originator's IP is
within the `10.200.0.0/30` CIDR range. Taking into account our example, only
`vm1`, `vm2`, and `vm3` will be able to contact any of its peers, as pictured
by the following diagram:

> ![MultiNetworkPolicy is provisioned](/assets/2023-07-10-OVN-kubernetes-secondary-networks-policies/02-no-access.png)

Reviewer: Please consider moving this explanation before the `MultiNetworkPolicy`, so it will be easier to understand what will happen before applying the manifest.

Author: Done.

---
layout: post
author: Miguel Duarte Barroso
title: NetworkPolicies for KubeVirt VMs secondary networks using OVN-Kubernetes
description: This post explains how to configure NetworkPolicies for KubeVirt VMs secondary networks.
navbar_active: Blogs

> pub-date: July 10

Author: Done.

but can target [network-attachment-definitions](https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/docs/how-to-use.md#create-network-attachment-definition).
[OVN-Kubernetes](https://github.com/ovn-org/ovn-kubernetes) implements this API, and configures access control accordingly
for secondary networks in the cluster.

In this post we will see how we can govern access control for VMs using the
multi-network policy API. On our simple example, we'll only allow into our VMs
for traffic ingressing from a particular CIDR range.

## Current limitations of `MultiNetworkPolicies` for VMs
Kubernetes `NetworkPolicy` has three types of policy peers:
- namespace selectors: allows ingress-from, egress-to based on the peer's namespace
- pod selectors: allows ingress-from, egress-to based on the peer's labels
- ip block: allows ingress-from, egress-to based on the peer's IP address

While `MultiNetworkPolicy` allows these three types, when used with VMs we
recommend using **only** the `IPBlock` policy peer - both `namespace` and `pod`
selectors prevent the live-migration of Virtual Machines (these policy peers

> require OVN-K managed IPAM, and currently the live-migration feature is only

Author: Your traffic will be dropped since the "allow" list IPs will be empty - OVN doesn't know what addresses are assigned to the logical ports.

available when IPAM is not enabled on the interfaces).

## Demo
To run this demo, we will prepare a Kubernetes cluster with the following
components installed:
- [OVN-Kubernetes](https://github.com/ovn-org/ovn-kubernetes)
- [multus-cni](https://github.com/k8snetworkplumbingwg/multus-cni)
- [KubeVirt](https://github.com/kubevirt/kubevirt)
- [Multi-Network policy API](https://github.com/k8snetworkplumbingwg/multi-networkpolicy)

The [following section](#setup-demo-environment) will show you how to create a
[KinD](https://kind.sigs.k8s.io/) cluster, with upstream latest OVN-Kubernetes,

> and upstream latest multus-cni deployed.

Author: Done.

kubectl apply -f "https://github.com/kubevirt/kubevirt/releases/download/${RELEASE}/kubevirt-operator.yaml"
kubectl apply -f "https://github.com/kubevirt/kubevirt/releases/download/${RELEASE}/kubevirt-cr.yaml"
kubectl -n kubevirt wait kv kubevirt --timeout=360s --for condition=Available
```

Now we have a Kubernetes cluster with all the pieces to start the Demo.

### Limiting ingress to a KubeVirt VM
In this example, we will configure a `MultiNetworkPolicy` allowing ingress into
our VMs only from a particular CIDR range - let's say `10.200.0.0/30`.

Provision the following NAD (to allow our VMs to live-migrate, we do not define
a `subnet`):
```yaml
---
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  name: flatl2net
spec:
  config: |2
    {
            "cniVersion": "0.4.0",
            "name": "flatl2net",
            "type": "ovn-k8s-cni-overlay",
            "topology":"layer2",
            "netAttachDefName": "default/flatl2net"
    }
```

Let's now provision our six VMs, with the following name to IP address
(statically configured via cloud-init) association:
- vm1: 10.200.0.1
- vm2: 10.200.0.2
- vm3: 10.200.0.3
- vm4: 10.200.0.4
- vm5: 10.200.0.5

> - vm6: 10.200.0.6.

Reviewer: nit:
```suggestion
- vm6: 10.200.0.6
```

[fedora@vm4 ~]$ ping 10.200.0.1 -c 4
PING 10.200.0.1 (10.200.0.1) 56(84) bytes of data.
64 bytes from 10.200.0.1: icmp_seq=1 ttl=64 time=3.20 ms
64 bytes from 10.200.0.1: icmp_seq=2 ttl=64 time=1.62 ms
64 bytes from 10.200.0.1: icmp_seq=3 ttl=64 time=1.44 ms
64 bytes from 10.200.0.1: icmp_seq=4 ttl=64 time=0.951 ms

--- 10.200.0.1 ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 3006ms
rtt min/avg/max/mdev = 0.951/1.803/3.201/0.843 ms
[fedora@vm4 ~]$ ping 10.200.0.6 -c 4
PING 10.200.0.6 (10.200.0.6) 56(84) bytes of data.
64 bytes from 10.200.0.6: icmp_seq=1 ttl=64 time=1.85 ms
64 bytes from 10.200.0.6: icmp_seq=2 ttl=64 time=1.02 ms
64 bytes from 10.200.0.6: icmp_seq=3 ttl=64 time=1.27 ms
64 bytes from 10.200.0.6: icmp_seq=4 ttl=64 time=0.970 ms

--- 10.200.0.6 ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 3005ms
rtt min/avg/max/mdev = 0.970/1.275/1.850/0.350 ms
```

We will now provision a `MultiNetworkPolicy` applying to all pods attached to
the `flatl2net` secondary network in the `default` namespace (from the
`k8s.v1.cni.cncf.io/policy-for` annotation) having the `access-control` label (

> from the network policy pod selector `matchLabels` stanza).

Reviewer 2: This is correct, but confusing since we never defined any Pods, just VMs.

Would you please use this to explain that this applies to both Pods and VMs. Where the `podSelector.matchLabels` matches labels set in VM's `spec.template.metadata.labels`?

Our users should not need to know that a VM is turned into a VMI which is then turned into a Pod.

Author: I agree with your conclusion, but I don't understand the rest. 

Is it better if I say `We will now provision a MultiNetworkPolicy applying to all VMs attached to` instead (while leaving the rest unchanged) ?

Reviewer 2: ```suggestion
We will now provision a `MultiNetworkPolicy` applying to all the VMs defined above.
To do this mapping correctly, the policy has to:

* Be in the same namespace as the VM.
* Set `k8s.v1.cni.cncf.io/policy-for` annotation matching the secondary network used by the VM.
* Set `matchLabels` selector matching the labels set on VM's `spec.template.metadata`.
```

Author: Done.

This policy will allow ingress into these `access-control` labeled pods 

> **only if** the traffic originates from within the `10.200.0.0/30` CIDR range.

Reviewer 2: ```suggestion
**only if** the traffic originates from within the `10.200.0.0/30` CIDR range
(IPs 10.200.0.1-3).
```

For the less network savvy.
